### PR TITLE
Issue #728 - Add support for address formats in different locales

### PR DESCRIPTION
**Pull request by AcharyaS97**
Minimal, initial fix for issue #728. We add an enum to the TransportNetwork to specify the output format (Street Name first vs Street Number First) and use that to set the format in the LocationAdapter, through a LocationFormatChanger object.

Idea was to intercept search results during the filter phase and use the LocationFormatChanger object to modify the format of the search result string.

### CODE REVIEW

**Comment by AcharyaS97** [comment]
I accept the CLA

**Comment by AcharyaS97** [comment]
I accept the CLA ✔️

**Comment by grote** [comment]
Thanks for the PR! The clabot will only shutup once the commit with the CLA acceptance has been merged to master unfortunately.

**Comment by AcharyaS97** [comment]
No worries. So what do you think about my set of changes?

File: app/src/main/java/de/grobox/transportr/locations/LocationAdapter.java [hunk 1]
```diff
@@ -218,4 +221,9 @@ void setSort(FavLocationType favLocationType) {
 		resetDropDownLocations();
 	}
 
+	void setLocationFormatChanger(TransportNetwork.LocationFormat format){
```
Comment (ialokim) [inline app/src/main/java/de/grobox/transportr/locations/LocationAdapter.java:224]: Naming: You're not setting the changer, but the locationFormat here.

File: app/src/main/java/de/grobox/transportr/locations/LocationFormatChanger.kt [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+        //Keep appending tokens until you get to the very last token that you already inserted
+        for (i in 0 until tokens.size - 1) {
+            str.append(tokens[i])
+            str.append(" ")
+        }
+
+        return str.toString().trim { it <= ' ' }
+    }
+    private fun isInteger(s: String): Boolean {
```
Comment (ialokim) [inline app/src/main/java/de/grobox/transportr/locations/LocationFormatChanger.kt:78]: Might be checked in a more idiomatic way using [`toIntOrNull`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/to-int-or-null.html).

**Review by ialokim — CHANGES_REQUESTED** [review]
Thanks for your PR!

This only influences the suggestions as displayed in the search right? Would it make sense to change formatting also at other places, e.g. in the trips view / work/home locations? (Or are addresses already correctly formatted there?) If not, I fear you might get inconsistent names throughout the app.

File: app/src/main/java/de/grobox/transportr/locations/SuggestLocationsFilter.java [hunk 3]
```diff
@@ -32,16 +33,16 @@
 	/**
 	 * Starts an asynchronous filtering operation on a worker thread.
 	 */
-	FilterResults performFiltering(CharSequence constraint, List<WrapLocation> favoriteLocations, List<SuggestedLocation> suggestedLocations) {
+	FilterResults performFiltering(CharSequence constraint, List<WrapLocation> favoriteLocations, List<SuggestedLocation> suggestedLocations, LocationFormatChanger locationFormatChanger) {
 		FilterResults filterResults = new FilterResults();
 		if (constraint == null) return filterResults;
-		List<WrapLocation> result = new ArrayList<>();
+		HashSet<WrapLocation> result = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (AcharyaS97) [reply app/src/main/java/de/grobox/transportr/locations/SuggestLocationsFilter.java:39]: Oh that's true. I could use a LinkedHashSet to maintain both order and uniqueness?

File: app/src/main/java/de/grobox/transportr/locations/LocationFormatChanger.kt (hunk 2 above)
Comment (AcharyaS97) [reply app/src/main/java/de/grobox/transportr/locations/LocationFormatChanger.kt:78]: Sure!

File: app/src/main/java/de/grobox/transportr/locations/LocationAdapter.java (hunk 1 above)
Comment (AcharyaS97) [reply app/src/main/java/de/grobox/transportr/locations/LocationAdapter.java:224]: Right. I will change the name

File: app/src/main/java/de/grobox/transportr/locations/SuggestLocationsFilter.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ialokim) [reply app/src/main/java/de/grobox/transportr/locations/SuggestLocationsFilter.java:39]: Sounds good to me.

**Comment by grote** [comment]
AFAIK this is a problem of Navitia and should be addressed there. When I met them some years back in Paris, they already said they will be launching a new search engine that handles different address format better. I wonder what become of this.

**Comment by AcharyaS97** [comment]
The reason I figured it would be alright leaving the updates out in other places was from the perspective that people wouldn't necessarily change TransportNetworks all too often and if they did, they wouldn't care about the format olocations they saved in other those places (ie work or home or favourite trips) . If you're based in an area where the address is in one style and your preferences are in that style then you go travelling and you switch transport networks, you wouldn't really care if it changed the locations you saved in other networks. I guess that leads into being able to set an address format as a setting though.

Also to the other point, I agree its an upstream issue but lieu of a solution there I figured it might improve usability of the app immediately for people in the jurisdictions of transport networks like mine where the upstream data format isn't the same as what the "accepted" format is.

**Comment by ialokim** [comment]
> The reason I figured it would be alright leaving the updates out in other places was from the perspective that people wouldn't necessarily change TransportNetworks all too often and if they did, they wouldn't care about the format olocations they saved in other those places (ie work or home or favourite trips) . If you're based in an area where the address is in one style and your preferences are in that style then you go travelling and you switch transport networks, you wouldn't really care if it changed the locations you saved in other networks. I guess that leads into being able to set an address format as a setting though.

You might have misunderstood what I meant: Even while using the same TransportNetwork, you have several places in the app where addresses might be shown - the search suggestion pop-up is only one of them. Others include addresses in routing results or the work/home location set for this TransportNetwork. I guess it would also be necessary to post-process the addresses there?

> Also to the other point, I agree its an upstream issue but lieu of a solution there I figured it might improve usability of the app immediately for people in the jurisdictions of transport networks like mine where the upstream data format isn't the same as what the "accepted" format is.

I agree that it makes sense to improve usability for users if the upstream is not going to fix it soon. Perhaps it might be good to ask them before, how their plans are concerning the new search engine. At last, this "hack" should be clearly marked as such and the respective functions should only be called for the specific TransportNetworks where necessary.

**Pull request closed without merging**